text/2978-stack_based_vec.md

```diff
@@ -0,0 +1,281 @@
+- Feature Name: `stack_based_vec`
+- Start Date: 2020-09-27
+- RFC PR: [rust-lang/rfcs#2990](https://github.com/rust-lang/rfcs/pull/2990)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+This RFC, which depends and takes advantage of the upcoming stabilization of constant generics (min_const_generics), tries to propose the creation of a new "growable" vector named `ArrayVec` that manages stack memory and can be seen as an alternative for the built-in structure that handles heap-allocated memory, aka `alloc::vec::Vec<T>`.
+
+# Motivation
+[motivation]: #motivation
+
+`core::collections::ArrayVec<T>` has several use-cases and should be conveniently added into the standard library due to its importance.
+
+### Unification
+
+There are a lot of different crates about the subject that tries to do roughly the same thing, a centralized implementation would stop the current fragmentation.
+
+### Optimization
+
+Stack-based allocation is generally faster than heap-based allocation and can be used as an optimization in places that otherwise would have to call an allocator. Some resource-constrained embedded devices can also benefit from it.
```

Comment:
This is using inline storage, so there is no indirection on nested storage, e.g. if we have an `ArrayVec` of `Arrayvec`s.
This is like the same argument for using `Vec<T>` rather than `LinkedList<T>`

Reply:
So you mean it is reserved during compile-time and thats why it should be treated special?

I did not hear an argument, why it will not be possible to "slap a const before allocator operations" (with extending the `Vec` API in some way to make it usable in core) to get the same effect. Maybe I do also have flawed assumptions.

Reply:
If you have a list of `ArrayVec<i32, 3>`, representing [[1, 2, 3], [4, 5, 6]], it is laid out in memory contiguously, like this:

```text
+---------------------+---------------------+
| +-----+---+---+---+ | +-----+---+---+---+ |
| | len | 1 | 2 | 3 | | | len | 4 | 5 | 6 | |
| +-----+---+---+---+ | +-----+---+---+---+ |
+---------------------+---------------------+
```

If you have a list of `Vec<i32>`, it is instead laid out, requiring indirection through a pointer, like this:

```text
+---------------------+---------------------+
| +-----+-----+-----+ | +-----+-----+-----+ |
| | ptr | len | cap | | | ptr | len | cap | |
| +--|--+-----+-----+ | +--|--+-----+-----+ |
+----|----------------+----|----------------+
     |                     |
     |   +---+---+---+     |   +---+---+---+
     '-> | 1 | 2 | 3 |     '-> | 4 | 5 | 6 |
         +---+---+---+         +---+---+---+
```

Reply:
That should be `ArrayVec<i32, 6>` and would make a nice explanation. Can you add the base address, capacity to the upper ascii graphic? Ideally, the invalid parts would be sketched as well.

I think my assumptions, how the stack is generated in the compiler is flawed.

Reply:
Like this then:

If you have a `Vec<ArrayVec<i32, 4>>`, representing [[1, 2, 3], [4, 5]], it is laid out in memory contiguously, apart from the pointer, like this:

```text
Vec<ArrayVec<i32, 4>>
 +-----+-----+-----+
 | ptr | len | cap |
 +--|--+-----+-----+
    |
    |   +------------------------------+--------------------------+----------+
    |   |       ArrayVec<i32, 4>       |     Arrayvec<i32, 4>     |          |
    |   | +-----+---+---+---+--------+ | +-----+---+---+--------+ |  Unused  |
    '-> | | len | 1 | 2 | 3 | Unused | | | len | 4 | 5 | Unused | | capacity |
        | +-----+---+---+---+--------+ | +-----+---+---+--------+ |          |
        +------------------------------+--------------------------+----------+
```

If you have a `Vec<Vec<i32>>`, it is instead laid out like this, requiring indirection through two pointers.

```text
Vec<Vec<i32>>
 +-----+-----+-----+
 | ptr | len | cap |
 +--|--+-----+-----+
    |
    |   +---------------------+---------------------+----------+
    |   |       Vec<i32>      |       Vec<i32>      |          |
    |   | +-----+-----+-----+ | +-----+-----+-----+ |  Unused  |
    '-> | | ptr | len | cap | | | ptr | len | cap | | capacity |
        | +--|--+-----+-----+ | +--|--+-----+-----+ |          |
        +----|----------------+----|----------------+----------+
             |                     |
             |                     |   +---+---+--------+
             |                     '-> | 4 | 5 | Unused |
             |                         +---+---+--------+
             |   +---+---+---+--------+
             '-> | 1 | 2 | 3 | Unused |
                 +---+---+---+--------+
```

Reply:
@mbartlett21 i have used your illustration in the RFC. Hope you don't mind it